Introduce "-internal" variant of bridging header import flags

DougGregor opened the pull request:
The flags "-import-bridging-header" and "-import-pch" import a bridging header, treating the contents as a `public` import. Introduce "internal-" variants of both flags that provide the same semantics, but treat the imported contents as if they came in through an `@_implementationOnly internal` import instead. This helps us keep the the bridging header as an implementation detail.

This is only plausibly safe with library evolution enabled, so warn about that.

Tracked by rdar://74011750.

DougGregor committed:
[Options] Make `-import-bridging-header` the canonical spelling instead of `-import-objc-header`

This command-line option hasn't been Objective-C specific ever, really.
Make the language-independent spelling the primary one to make that
more obvious.

DougGregor committed:
Introduce "-internal" variant of bridging header import flags

The flags "-import-bridging-header" and "-import-pch" import a bridging
header, treating the contents as a public import. Introduce
"internal-" variants of both flags that provide the same semantics,
but are intended to treat the imported contents as if they came in
through an internal import. This is just plumbing of the options for
the moment.

DougGregor committed:
Treat internally-imported bridging headers as internally-imported

When using an internal import for a bridging header, semantically treat
the contents of the bridging header, and anything that it imports, as
if they were imported internally. This is the actual semantic behavior
we wanted from internally-imported bridging headers.

This is the main semantic checking bit for rdar://74011750.

tshortli commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2703,8 +2703,8 @@ NOTE(decl_import_via_here,none,
      "'%select{private|fileprivate|internal|package|%ERROR|%ERROR}1' "
-     "from %2",
-     (const Decl *, AccessLevel, const ModuleDecl*))
+     "from %select{%2|bridging header}3",
```
I feel like there may be more resilience diagnostics that will need to be updated, but it seems like this is a good start

DougGregor replied:
Perhaps! They'll come out as the weird __ObjC module if we don't customize them.

DougGregor replied:
Ah, that could help here, yes.

DougGregor replied:
I did this in https://github.com/swiftlang/swift/pull/84410/commits/b7a83495fa58b927f05346161ac67760f5356779

DougGregor committed:
Exempt C++ namespaces from internal-import checking

C++ namespaces always get imported into the bridging header's module
for Reasons. Exempt them from internal-import checking, since we get
checking for the namespace members that are actually used.

DougGregor committed:
Warn about the use of internal bridging headers without library evolution

It's very, very easy to make a mistake that will cause broken
serialized modules. Until that's no longer true, at least tell folks
that they are heading into uncharted waters, as we do with
`@_implementationOnly` imports.

xymus commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2703,8 +2703,8 @@ NOTE(decl_import_via_here,none,
-     "from %2",
-     (const Decl *, AccessLevel, const ModuleDecl*))
+     "from %select{%2|bridging header}3",
+     (const Decl *, AccessLevel, const ModuleDecl*, bool))
```
A similar precision would be a nice to have for the exportability diagnostics referencing `DisallowedOriginKind`.

xymus commented on lib/Sema/ImportResolution.cpp:
```diff
@@ -574,8 +574,14 @@ ModuleImplicitImportsRequest::evaluate(Evaluator &evaluator,
         ImportedModule(headerModule), SourceLoc(), ImportFlags::Exported);
+    if (ctx.ClangImporterOpts.BridgingHeaderIsInternal) {
+      import.accessLevel = AccessLevel::Internal;
+      import.options |= ImportFlags::ImplementationOnly;
```
The mix of `@_implementationOnly internal import` isn't exactly an expected configuration at the moment. I'm worried that checks for both could overlap and duplicate the exportability diagnostics. We could use only `internal import` for library-evolution enabled modules if it's an issue.

Do we still see the warnings on implementation-only use from non-library-evolution modules for an implicit import like this?

DougGregor committed:
[Serialization] Don't serialize an internally-imported bridging header

Internally-imported bridging headers must not leak outside of the Swift
module. Don't serialize their contents, and make sure we can still
import the module even if the bridging header has been removed.

DougGregor committed:
Improve some diagnostic with @_implementationOnly violations from internal bridging headers

DougGregor committed:
Don't mix internal and @_implementationOnly

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci clean smoke test macOS

DougGregor committed:
Add #include guards around bridging header

DougGregor commented:
@swift-ci clean smoke test macOS

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test Windows

DougGregor committed:
Expand DisallowedOriginKind with an explicit entry for internal bridging headers

DougGregor committed:
Add -internal-import-bridging-header as a supported feature

DougGregor committed:
Ensure that -internal-import-bridging-header doesn't show up in a textual interface

DougGregor committed:
Temporarily disable on Windows

DougGregor commented:
@swift-ci please smoke test

ian-twilightcoder commented on include/swift/Frontend/FrontendOptions.h:
```diff
@@ -66,12 +66,16 @@ class FrontendOptions {
 
-  /// An Objective-C header to import and make implicitly visible.
+  /// A C header to import and make implicitly visible.
   std::string ImplicitObjCHeaderPath;
```
Should we rename this and the one below it? Or save that for another day?

ian-twilightcoder commented on include/swift/Option/Options.td:
```diff
@@ -347,16 +347,25 @@ def import_underlying_module : Flag<["-"], "import-underlying-module">,
+  HelpText<"Implicitly imports a C header file">;
+def import_objc_header : Separate<["-"], "import-objc-header">,
   Flags<[FrontendOption, HelpHidden, ArgumentIsPath]>,
-  Alias<import_objc_header>;
```
Should we make swift-build switch to `-import-bridging-header`? How long has that been available?

ian-twilightcoder commented on include/swift/Option/Options.td:
```diff
@@ -347,16 +347,25 @@ def import_underlying_module : Flag<["-"], "import-underlying-module">,
+
+def internal_import_bridging_header : Separate<["-"], "internal-import-bridging-header">,
+  Flags<[FrontendOption, ArgumentIsPath]>,
+  HelpText<"Imports an C header file as an internal import">;
```
Implicitly internally imports a C header file? (Reads a little awkward but matches `import-bridging-header` a little closer. At the least "an C header" -> "a C header".)

ian-twilightcoder commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -3762,8 +3763,14 @@ ImportDecl *swift::createImportDecl(ASTContext &Ctx,
+    ID->getAttrs().add(
+        new (Ctx) AccessControlAttr(SourceLoc(), SourceRange(),
+                                    AccessLevel::Internal, /*implicit=*/true));
+    ID->getAttrs().add(new (Ctx) ImplementationOnlyAttr(/*IsImplicit=*/true));
```
Why do we want ImplementationOnlyAttr in addition to AccessLevel::Internal?

ian-twilightcoder commented on lib/Sema/TypeCheckAccess.cpp:
```diff
@@ -2692,6 +2699,11 @@ void swift::recordRequiredImportAccessLevelForDecl(
 
+  // Egregious hack: if the declaration is for a C++ namespace, assume it's
+  // accessible.
+  if (isa_and_nonnull<clang::NamespaceDecl>(decl->getClangDecl()))
```
Was this meant to be part of this change?

ian-twilightcoder commented on lib/Sema/TypeCheckAccess.h:
```diff
@@ -47,6 +47,10 @@ enum class DisallowedOriginKind : uint8_t {
   FragileCxxAPI,
+
+  /// An import that is internal via the internally-imported bridging header.
+  InternalBridgingHeaderImport,
```
Would this be more consistent as NonPublicBridgingHeaderImport?

ian-twilightcoder reviewed (approved):
This is awesome, thank you so much for doing this!

DougGregor committed:
Remove unnecessary addition of @_implementationOnly

DougGregor committed:
Clean up help text for new option
